nucleus/payara-modules/requesttracing-core/src/main/java/fish/payara/nucleus/requesttracing/store/LocalRequestTraceStore.java

```diff
@@ -51,29 +54,21 @@
  */
 public class LocalRequestTraceStore implements RequestTraceStoreInterface {
 
-    private BoundedTreeSet<RequestTrace> store;
+    private final Set<RequestTrace> store = ConcurrentHashMap.newKeySet();
```

Comment:
Since the strategy will already prevent the store from growing to large there is no need to also have a bound data structure. More importantly the concurrent hash set will have the expected behaviour. `BoundedTreeSet` will not work. Try this test:

```
@Test
public void test() {
  Set<RequestTrace> set = new ConcurrentSkipListSet<>(); // base class of BoundedTreeSet already show the problem
  for (int i = 0; i < 10; i++) {
    set.add(createRandomRequestTraceWithSingleSpanAndSameTime());
  }
  assertEquals(10, set.size());
}
```
The set should have size 10 but the `ConcurrentSkipListSet` actually can return any number between 1 and 10 at this point.

> Beware that, unlike in most collections, the size method is not a constant-time operation. Because of the asynchronous nature of these sets, determining the current number of elements requires a traversal of the elements, and so may report inaccurate results if this collection is modified during traversal.

This is just a hint the javadoc gives us on how the class works. Together with the improper implementation of `RequestTrace.compareTo` the size does come out wrong most of the times. 
I here opted to simply switch the collection as hash set is more adequate anyway. Still we do need to fix the `RequestTrace.compareTo` method but since that might have far reaching consequences I didn't want to bite this bullet in this PR.

Reply:
Is the functionality the same though?
The bounded tree set was used specifically so that the last (read: shortest) trace would be replaced if the store was full (assuming you're using the longest traces strategy)

Reply:
Yes and no. 

No: The property of the tree set to remove the shortest on "overflow" after adding is practically superfluous. If the strategy works correctly the strategy already will determine the element to remove before it ever comes to an overflow as the overflow occurs at max-size + 2 but strategy will remove max-size +1 to max-size after an add.  

Yes: If the strategy or other code involved to remove elements on behalf of the strategy does not work correctly (as it was) the "overflow" logic might have kick in and removed the shortest trace. I say might because of the `size()` bug this wasn't guaranteed either. Also this might be wrong in case the other strategy is used. Since there now are tests that verify correct behaviour I think we don't need to rely on the collection rescuing the failing strategy. I guess this behaviour has contributed to hiding the issues in the rest of the involved code.